### DISCUSSION TITLE
Resource is not binded when updating dependent fields

### DISCUSSION THREAD

**Discussion by xel1045** [laravel/nova-issues#4842] [Ideas & Feature Requests] [answered]
- Laravel Version: 9.27.0
- Nova Version: 4.14.3
- PHP Version: 8.1.9
- Database Driver & Version: MySQL 8.0.30
- Operating System and Version: macOS 12.6
- Browser type and version: Chrome 105

### Description:

When updating an existing resource, Laravel doesn't bind the resource when applying dependent fields.

To be more specific, when the `fields` method is call for a request to `/nova-api/resource/id/update-fields`, the `$this->resource` model is empty. In my case, it looks like this:

![Screen Shot 2022-09-23 at 11 46 28](https://user-images.githubusercontent.com/1497697/192000799-62c814e4-0899-46a7-adf3-d29c87672000.png)

It seems that `\Laravel\Nova\Http\Controllers\UpdateFieldController::sync` was modified in Nova 4.2.4 and it doesn't use the binded resource.

Right now it looks like this:

```php
    public function sync(ResourceUpdateOrUpdateAttachedRequest $request)
    {
        UpdateViewResource::make()->newResourceWith($request);

        return response()->json(
            $request->newResource()
                ->updateFields($request)
```

But I think it should be this instead:

```php
    public function sync(ResourceUpdateOrUpdateAttachedRequest $request)
    {
        $resource = UpdateViewResource::make()->newResourceWith($request);

        return response()->json(
            $resource
                ->updateFields($request)
```

### Detailed steps to reproduce the issue on a fresh Nova installation:

1. Create a resources with dependents fields
2. Add a breakpoint on the `fields` method when `update-fields` on edit (condition `str_contains($request->url(), 'update-fields')`)
3. Change a dependent field to trigger an update
4. When your breakpoint is reached, check the current value of the `$this->resource`

Thanks!

**Comment by crynobone** [comment]
This is the current expected behaviour, as loading the resources on each sync could be a performance bottleneck for the app. Marking this as a feature request for now.

**Reply by xel1045** [reply]
Indeed, it's important that this endpoint stay performant! However, since the `UpdateViewResource` class is already loading the resource to perform the authorization, why don't reuse the resource it creates?

The downside of this is that we cannot have any readonly, required, rules, etc. based on the loaded resource, because it will be overwritten when the field is updated.

**Reply by crynobone** [reply]
What would be a use-case for you to get the current resource from database?

**Reply by xel1045** [reply]
We have a "Province / State" select field which depends on the selected country and is readonly on edit if the model was manually created through the admin. Unfortunately, the readonly state is lost when options are loaded.

**Reply by crynobone** [reply]
Can you provide a code example for the use case?

**Reply by xel1045** [reply]
Unfortunately, I don't have a public repository for this, but the code looks like this:

```php
class Donation {
    public function fields(): array
    {
        return [
            // ...

            Select::make('State', 'state_id')
                    ->dependsOn('country_id', function (Select $field, NovaRequest $request, FormData $formData) {
                        if (! $countryId = $formData->get('country_id')) {
                            return;
                        }

                        $country = Country::query()->findOrFail($countryId);
                        $states = $country->states()->pluck('name', 'id');

                        $field
                            ->options($states)
                            ->readonly(! $this->resource->exists || $this->resource->is_manually_created)
                            ->rules([
                                'required',
                                'integer',
                                Rule::in($states->keys()),
                            ]);
                    })
                    ->displayUsingLabels()
                    ->readonly(),

            // ...
        ];
    }
}
```

Right now, our workaround is to load the resource when it's not loaded in update requests, but this loads the model twice (in `\Laravel\Nova\Http\Controllers\UpdateFieldController::sync` and in our method `isManuallyCreated`) and this is really hacky:

```php
    /**
     * @param \Laravel\Nova\Http\Requests\NovaRequest $request
     * @return bool
     */
    protected function isManuallyCreated(NovaRequest $request): bool
    {
        $resource = $this->resource;

        // When creating a donation, we can assume it's manually created
        if ($request->isUpdateOrUpdateAttachedRequest()) {
            $resource = once(fn () => $resource->exists ? $resource : $request->findModel());
        }

        return ! $resource->exists || $resource->is_manually_created;
    }
```
As mentioned, it was working prior to Nova 4.2.4.

Thanks!

**Reply by crynobone** [reply]
Personally I would use:

```php
class Donation {
    public function fields(): array
    {
        return [
            // ...
            Hidden::make('Is Manually Created', 'is_manually_created'),
            Select::make('State', 'state_id')
                    ->dependsOn(['country_id', 'is_manually_created'], function (Select $field, NovaRequest $request, FormData $formData) {
                        if (! $countryId = $formData->get('country_id')) {
                            return;
                        }

                        $country = Country::query()->findOrFail($countryId);
                        $states = $country->states()->pluck('name', 'id');

                        $field
                            ->options($states)
                            ->readonly(! $formData->is_manually_created)
                            ->rules([
                                'required',
                                'integer',
                                Rule::in($states->keys()),
                            ]);
                    })
                    ->displayUsingLabels()
                    ->readonly(),

            // ...
        ];
    }
}
```

**Reply by xel1045** [reply]
Indeed, that would work too as a workaround.

**Comment by crynobone** [comment, marked as answer]
Implemented in 4.16

**Reply by xel1045** [reply]
Thanks!
